### ISSUE TITLE
Add keyframe-selector-notation

### ISSUE THREAD

**Issue by MartijnCuppens** [stylelint/stylelint#4065]
### What is the problem you're trying to solve?

In animations the as well the `from/to` syntax as the `0%/100%` syntax can be used.

For example:
```css
@keyframes animation-1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}

@keyframes animation-2 {
  0% {
    color: green;
  }
  100% {
    color: yellow;
  }
}
```

### What solution would you like to see?

A new rule that would determine which approach you should use. Not sure how this should be implemented, because sometimes you'll have an animation which looks like this:

```css
@keyframes animation-2 {
  20% {
    color: green;
  }
  100% {
    color: yellow;
  }
}
```
Which will lead to a mix of `20%` and `to` if the `from/to` syntax is preferred.

**Comment by hudochenkov** [comment]
Thank you for a proposal. I think it could be in stylelint to enforce consistency. I propose the following:

- Name: `keyframe-selector-keyword`
- Primary option: `"always"|"never"`
- Secondary options: none
- Messages: 
  - "Expected keyframe selector to be a keyword “from/to”"
  - "Expected keyframe selector to be a 0%/100%"
- Description: "Require or disallow keywords as selector for keyframe rules."
- Section: "Limit language features" -> "Keyframes"

@stylelint/contributors what do you think?

[Keyframes — CSS Animations Level 1](https://drafts.csswg.org/css-animations/#keyframes)
[@keyframes - CSS: Cascading Style Sheets | MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes)

**Comment by b8kkyn** [comment]
For me it's totally fine except one thing.

Rule should not force only 0% and 100% values. It should be any percent value between 0 and 100.

So we can say,
_"Expected keyframe selector to be a percent"_ **not** _"Expected keyframe selector to be a 0%/100%"_

I believe that the person who asked the question is tried to express this.

**Comment by hudochenkov** [comment]
Well, if it's not percent then it's only `from` or `to`, which are `0%` and `100%`. They couldn't be any other percentage.

**Comment by b8kkyn** [comment]
Sorry, I'm not quite sure understand but I believe this proposal is all about having rule to determine which approach should be used. And these approaches are "from/to" or "percent value" not only "0%/100%". I tried to say this.

**Comment by hudochenkov** [comment]
Correct. But if keywords were used, when rule was set  to accept percentages only, we know what percentage equivalent of keyword is and we can tell it to user and autofix.

**Comment by b8kkyn** [comment]
Yes we can do that. We know equivalent of `from` and `to` which are `0%` and `100%`. But vice versa cannot be done. When rule was set to accept only keywords and we had something like this:

```
@keyframes animation-2 {
  0& {
    color: purple;
  }
  20% {
    color: green;
  }
  100% {
    color: yellow;
  }
}
```

So what is `from`, what is `to`?

But from beginning I didn't talk about this :) I tried to say:

From message _"Expected keyframe selector to be a 0%/100%"_, it turns out like we can only use 0 and 100 percent values.

I say above example I gave should be valid if rule was set to only use percents.

**Comment by spiltcoffee** [comment]
Could the values be `prefer-percentage`/`prefer-keyword` instead, or something along those lines? 

`always`/`never` confuses me because if I was to make it `always`, does that mean I could never use `20%` as a step?

**Comment by hudochenkov** [comment]
> does that mean I could never use 20% as a step

Correct. If rule enforce keywords, then it won't be possible to use percentages without warnings or disabling a rule. And it doesn't matter what option name it would be.

**Comment by evilebottnawi** [comment]
Maybe `keyframe-selector-style: 'percentrage' | 'keyword'`?

**Comment by jeddy3** [comment]
Revisiting as 2022 issue clean-up...

A few of our conventions have solidified since this issue was opened. Let's merge @hudochenkov and @alexander-akait suggestions into:

- Name: `keyframe-selector-notation`
- Primary option: `"keyword"|"percentage"`
- Secondary options: none
- Autofixable: Yes
- Message: - `Expected "${selector}" to be a "${fixedSelector}" `
- Description: "Specify keyword or percentage notation for keyframe selectors."
- Extended description: "The keyword `from` is equivalent to the value `0%`. The keyword `to` is equivalent to the value `100%`."
- Section: "Enforce convention" -> "Keyframe selectors"

The rule only enforces the convention for `to`, `from`, `0%` and `100%`.

Refs:

- https://stylelint.io/user-guide/rules/list/alpha-value-notation - `*-notation` example
- https://stylelint.io/user-guide/rules/about#names - "explicit names" docs

[From spec](https://www.w3.org/TR/css-animations-1/#keyframes) - rationale for rule & option names, and scope of rule:

> The <keyframe-selector> for a <keyframe-block> consists of a comma-separated list of percentage values or the keywords from or to. The selector is used to specify the percentage along the duration of the animation that the keyframe represents. The keyframe itself is specified by the block of property values declared on the selector. _The keyword from is equivalent to the value 0%. The keyword to is equivalent to the value 100%._

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by Mouvedia** [comment]
There are browser bugs related to quoted/unquoted animation names.
Will we also have an animation-name-notation rule ?

**Comment by jeddy3** [comment]
> Will we also have an animation-name-notation rule ?

@Mouvedia Yes, we can add a `keyframes-name-quotes` rule. Please [open a new issue](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REQUEST_A_FEATURE.yml) so that we can spec it out.
